### ISSUE TITLE
Variable importance on a validation set, by tree

### ISSUE THREAD

**Issue by sebconort** [dmlc/xgboost#3088]
Hello,

I cheked but did not find any issue about computing variable importance on a validation or hold out set, by computing the gains on loss on that set.

That would indeed be very useful to get variable importance not based on over optimistic gains computed on train set. 

Also it would be nice that it be not approximated gains based on first and second order derivatives (I understand that it is computationally efficient to use derivates on the train set since those quantities are already computed to grow the trees, but on a validation set computing directly real gains would make more sense to me)

Also, having gains by tree and by variable would be great, so as to be able to graph the importance of each variable regarding the number of trees added. We could then better understand at what stage certain features are introduced, with what importance.

Best regards, and thanks to all contributors

**Comment by khotilov** [comment]
> computing variable importance on a validation or hold out set, by computing the gains on loss on that set.

See #1670 and also https://github.com/dmlc/xgboost/issues/2970#issuecomment-354684604 for some extra explanation.

> it would be nice that it be not approximated gains based on first and second order derivatives

I'm not sure about the benefits of "real" gains, as they are not exactly what is optimized to find splits. But if you are really interested, you may take the https://github.com/dmlc/xgboost/blob/master/src/tree/updater_refresh.cc as a prototype for your custom updater to compute and store the real gains.

> Also, having gains by tree and by variable would be great

In R, `xgb.model.dt.tree()` would give you a data.table that you may aggregate in whatever way you like. In Python, you'd have to parse the model dump on your own, but there are plenty of examples out there if you search around.

**Comment by sebconort** [comment]
Thanks a lot for this very instructive answer! I had not knowledge of the update and refresh possibilties on a new dataset. 
This answers my need of getting gains related to the optimised loss on a different dataset. I am going to try that!

But here are my thoughts on additional possibilities that could be very nice in the library.
When training a model, we can pass (eval_set, eval_metric) in sklearn API for instance. This way we get the value of the eval_metric on eval_set after each additional tree. What is nice is that eval_metric can be different from the optimised metric, and eval_set different from the training set. When all these computation are made, maybe it could be nice to compute get the gain contribution of each variable used in each tree?
Happy to get your feedback and expertise.
